source/random/engine/pcg.d

```diff
@@ -0,0 +1,625 @@
+                acc_mult *= _mult;
+                acc_plus = cast(Uint)(acc_plus * _mult + _inc);
+            }
+            _inc  *= _mult + 1;
+            _mult *= _mult;
+            delta >>= 1;
+        }
+        state = cast(Uint)(acc_mult*state + acc_plus);
+    }
+}
+
+/// 32-bit output PCGs with 64 bits of state.
+alias pcg32        = PermutedCongruentialEngine!(xsh_rr!(uint,ulong),stream_t.specific,true);
+///ditto
+alias pcg32_unique = PermutedCongruentialEngine!(xsh_rr!(uint,ulong),stream_t.unique,true);
+///ditto
+alias pcg32_oneseq = PermutedCongruentialEngine!(xsh_rr!(uint,ulong),stream_t.oneseq,true);
+///ditto
+alias pcg32_fast   = PermutedCongruentialEngine!(xsh_rr!(uint,ulong),stream_t.none,true);
+
+//alias pcg64        = PermutedCongruentialEngine!(xsh_rr!(ulong,ucent),stream_t.specific,true);
+//alias pcg64_unique = PermutedCongruentialEngine!(xsh_rr!(ulong,ucent),stream_t.unique,true);
+//alias pcg64_oneseq = PermutedCongruentialEngine!(xsh_rr!(ulong,ucent),stream_t.oneseq,true);
+//alias pcg64_fast   = PermutedCongruentialEngine!(xsh_rr!(ulong,ucent),stream_t.none,true);
+
+/// PCGs with n bits output and n bits of state.
+alias pcg8_once_insecure  = PermutedCongruentialEngine!(rxs_m_xs_forward!(ubyte ,ubyte ),stream_t.specific,true);
+///ditto
+alias pcg16_once_insecure = PermutedCongruentialEngine!(rxs_m_xs_forward!(ushort,ushort),stream_t.specific,true);
+///ditto
+alias pcg32_once_insecure = PermutedCongruentialEngine!(rxs_m_xs_forward!(uint  ,uint  ),stream_t.specific,true);
+///ditto
+alias pcg64_once_insecure = PermutedCongruentialEngine!(rxs_m_xs_forward!(ulong,ulong  ),stream_t.specific,true);
+//alias pcg128_once_insecure = PermutedCongruentialEngine!(rxs_m_xs_forward!(ucent,ucent,stream_t.specific,true);
+
+/// As above but the increment is not dynamically setable.
+alias pcg8_oneseq_once_insecure
+        = PermutedCongruentialEngine!(rxs_m_xs_forward!(ubyte ,ubyte ),stream_t.oneseq,true);
+///ditto
+alias pcg16_oneseq_once_insecure
+        = PermutedCongruentialEngine!(rxs_m_xs_forward!(ushort,ushort),stream_t.oneseq,true);
+///ditto
+alias pcg32_oneseq_once_insecure
+        = PermutedCongruentialEngine!(rxs_m_xs_forward!(uint  ,uint  ),stream_t.oneseq,true);
+///ditto
+alias pcg64_oneseq_once_insecure
+        = PermutedCongruentialEngine!(rxs_m_xs_forward!(ulong,ulong  ),stream_t.oneseq,true);
+//alias pcg128_oneseq_once_insecure
+//      = PermutedCongruentialEngine!(rxs_m_xs_forward!(ucent,ucent,stream_t.specific,true);
+
+private alias AliasSeq(T...) = T;
+
+@safe unittest
+{
+    foreach(RNG; AliasSeq!(pcg32,pcg32_unique,pcg32_oneseq,pcg32_fast,
+                           pcg8_once_insecure,pcg16_once_insecure,pcg32_once_insecure,pcg64_once_insecure,
+                           pcg8_oneseq_once_insecure,pcg16_oneseq_once_insecure,pcg32_oneseq_once_insecure,
+                           pcg64_oneseq_once_insecure))
+    {
+        auto gen = RNG(cast(RNG.Uint)unpredictableSeed);
+        gen();
```

Comment:
Should be doable for all but the unique types (since I don't know the address ahead of time). Might take me a bit of time as I start uni again tomorrow.